pkg/build/sbom.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2022 Chainguard, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	osr "github.com/dominodatalab/os-release"
+	"github.com/pkg/errors"
+	"gitlab.alpinelinux.org/alpine/go/pkg/repository"
+)
+
+// TODO(kaniini): Move most of this over to gitlab.alpinelinux.org/alpine/go.
+type document struct {
+	BOMFormat    string       `json:"bomFormat"`
+	SpecVersion  string       `json:"specVersion"`
+	Version      int          `json:"version"`
+	Components   []component  `json:"components,omitempty"`
+	Dependencies []dependency `json:"dependencies,omitempty"`
+}
+
+type component struct {
+	BOMRef             string              `json:"bom-ref"`
+	Type               string              `json:"type"`
+	Name               string              `json:"name"`
+	Version            string              `json:"version"`
+	Description        string              `json:"description"`
+	PUrl               string              `json:"purl"`
+	ExternalReferences []externalReference `json:"externalReferences,omitempty"`
+	Licenses           []license           `json:"licenses,omitempty"`
+	Components         []component         `json:"components,omitempty"`
+}
+
+type license struct {
+	Expression string `json:"expression"`
+}
+
+type externalReference struct {
+	URL  string `json:"url"`
+	Type string `json:"type"`
+}
+
+type dependency struct {
+	Ref       string   `json:"ref"`
+	DependsOn []string `json:"dependsOn"`
+}
+
+func bomRef(ns string, pkg *repository.Package) string {
+	return fmt.Sprintf("pkg:apk/%s/%s", ns, pkg.Name)
+}
+
+func bomPurl(ns string, pkg *repository.Package) string {
+	return fmt.Sprintf("pkg:apk/%s/%s@%s", ns, pkg.Name, pkg.Version)
+}
+
+func (bc *Context) GenerateSBOM() error {
```

Comment:
In case it is useful, @imjasonh rolled the `ko` CycloneDX stuff here if you want to compare or copy useful bits: https://github.com/google/ko/pull/587/files#diff-ad63c642b90be2eed057117c368556d45c8c89a4c7310c948b80166fed73667aR41

Reply:
Whether or not you roll your own, I think it was a good idea to have e2e validation of the SBOMs in CI: https://github.com/google/ko/blob/main/.github/workflows/sbom.yaml

Reply:
Also: how do folks feel about generating both/all formats by default? That's something I'd like to do for ko, and apko being onboard would help motivate me. Seems weird to prefer one over the other(s)

Reply:
I would be alright with generating the JSON-LD version of SPDX, but not the Turtle version presently generated by `ko`.

Reply:
(I also think that shouldn't block this MR)

Reply:
Absolutely, we should do both. I was holding on commenting on this one to talk on Monday (@kaniini happy to chat on the weekend too if you want). I think we need to align some things to work together, buts lets chat first!

Reply:
It can wait until Monday, I have some reasons for being opinionated here on SBOMs that I can outline outside this MR.

Reply:
After a discussion with @puerco, the plan is for him to take over this branch and add in the SPDX support and so on :)